boto3/docs/utils.py

```diff
@@ -11,11 +11,23 @@
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
 import inspect
+import re
 from botocore.compat import six
 
+TARGET_COMPONENT_RE = re.compile(r'[^.\[\]]+(?![^\[]*\])')
+
 
 def get_resource_ignore_params(params):
```

Comment:
`"foo[]bar"[].baz`.  Basically this is assuming that these characters in the regex cannot appear anywhere except in the places we want to ignore them (i.e `foo[]`).  This is likely ok for most, if not all the cases we'll ever need to consider but is not 100% correct.

This function is basically trying to parse a JMESPath expression and select out the first field name using a regex.  Rather than use this regex based approach which has edge cases and isn't 100% correct, why not using the existing JMESPath parser and not have any edge cases?